File: Scripts/make-dmg-background.py
#!/usr/bin/env python3
"""Draws the disk image's background.

    Scripts/make-dmg-background.py <output.png>

Generated rather than stored, for the same reason the app icon is: a public
repository should not carry a binary it can produce in a second, and a
generated image can be adjusted without a round trip through a design tool.

The geometry is tied to `build-dmg.sh`. That script opens a 660x400 window and
puts the app at (145, 185) and the Applications alias at (515, 185); the arrow
below is drawn between those two points, and the caption sits under them. If
either moves, both files move together.
"""
import math
import pathlib
import struct
import sys
import zlib

# Points; the window build-dmg.sh asks Finder for.
W, H = 660, 400
SCALE = 2                      # drawn at 2x so it is not soft on a Retina display
APP_AT = (145, 185)
APPS_AT = (515, 185)

PX_W, PX_H = W * SCALE, H * SCALE
pixels = [[(0, 0, 0, 255)] * PX_W for _ in range(PX_H)]


def put(x, y, colour, alpha=1.0):
    if not (0 <= x < PX_W and 0 <= y < PX_H) or alpha <= 0:
        return
    r0, g0, b0, _ = pixels[y][x]
    r, g, b = colour
    pixels[y][x] = (int(r0 + (r - r0) * alpha),
                    int(g0 + (g - g0) * alpha),
                    int(b0 + (b - b0) * alpha), 255)


# The ground: the same deep indigo the icon sits on, lifting slightly toward
# the bottom so the window has a horizon rather than being a flat field.
for y in range(PX_H):
    t = y / PX_H
    row = (int(14 + 10 * t), int(16 + 12 * t), int(26 + 18 * t))
    for x in range(PX_W):
        pixels[y][x] = row + (255,)

# A wide, very soft glow behind where the app icon lands, so the eye starts
# there rather than in the middle of the window.
gx, gy = APP_AT[0] * SCALE, APP_AT[1] * SCALE
radius = 150 * SCALE
for y in range(max(0, gy - radius), min(PX_H, gy + radius)):
    for x in range(max(0, gx - radius), min(PX_W, gx + radius)):
        d = math.hypot(x - gx, y - gy) / radius
        if d < 1:
            put(x, y, (255, 162, 60), 0.055 * (1 - d) ** 2)


def dot(cx, cy, r, colour, alpha):
    for y in range(int(cy - r) - 1, int(cy + r) + 2):
        for x in range(int(cx - r) - 1, int(cx + r) + 2):
            d = r - math.hypot(x - cx, y - cy)
            if d >= 0:
                put(x, y, colour, alpha * min(1.0, d))


# The arrow: a dotted run from the app to the Applications alias, fading in
# and out so it reads as a hint rather than an instruction. Drawn below the
# icons, which sit at y=185 and are 100 points tall.
y_line = (APP_AT[1] + 78) * SCALE
x0, x1 = (APP_AT[0] + 62) * SCALE, (APPS_AT[0] - 62) * SCALE
count = 13
for i in range(count):
    t = (i + 0.5) / count
    x = x0 + (x1 - x0) * t
    # Brightest in the middle; the ends are where the icons already are.
    fade = math.sin(math.pi * t) ** 0.7
    dot(x, y_line, 2.6 * SCALE, (255, 186, 105), 0.5 * fade)

# The head, a simple triangle so nothing here needs a font.
#
# Apex on the *right*, toward Applications. The first version tapered the
# other way and pointed back at the app — which tells somebody to drag
# Applications onto Lathe, the exact opposite of the instruction.
hx = x1 + 6 * SCALE
length = int(11 * SCALE)
for i in range(length):
    reach = length - i
    for j in range(-reach, reach + 1):
        put(hx + i, y_line + j, (255, 186, 105),
            0.55 * (1 - i / length) ** 0.5)

raw = b"".join(
    b"\x00" + bytes(v for px in row for v in px)
    for row in pixels)


def chunk(tag, data):
    return (struct.pack(">I", len(data)) + tag + data
            + struct.pack(">I", zlib.crc32(tag + data) & 0xFFFFFFFF))


out = pathlib.Path(sys.argv[1] if len(sys.argv) > 1 else "dmg-background.png")
out.parent.mkdir(parents=True, exist_ok=True)
out.write_bytes(
    b"\x89PNG\r\n\x1a\n"
    + chunk(b"IHDR", struct.pack(">IIBBBBB", PX_W, PX_H, 8, 6, 0, 0, 0))
    # A physical-pixel chunk so Finder renders it at 1x window size on a
    # Retina display rather than doubling it.
    + chunk(b"pHYs", struct.pack(">IIB", 2835 * SCALE, 2835 * SCALE, 1))
    + chunk(b"IDAT", zlib.compress(raw, 9))
    + chunk(b"IEND", b""))
print(f"    background: {out} ({PX_W}x{PX_H})")

File: Scripts/make-dmg-background.swift
#!/usr/bin/env swift
// Draws the disk image's background.
//
//     swift Scripts/make-dmg-background.swift <output.png>
//
// Generated rather than stored, for the same reason the app icon is: a public
// repository should not carry a binary it can produce in a second.
//
// ## Why light
//
// The previous background was near-black, and that was a defect rather than a
// style. Finder draws the labels under the icons in DARK text whatever the
// background image is — it does not look at the picture — so on a dark ground
// "Lathe" and "Applications" were close to invisible. A light ground is the
// only choice that keeps Finder's own text legible, and it matches the
// documentation site, which is a light steel spec sheet.
//
// ## Why Swift
//
// The old generator was a hand-rolled rasteriser in pure Python, which could
// draw dots and a triangle and nothing that needed a font. Everything this
// wants to say is text. CoreGraphics and CoreText are on every Mac, including
// the release runner, so there is no dependency to add.
//
// ## Geometry
//
// Tied to `build-dmg.sh`: a 660x400 window, 100pt icons, the app centred at
// (145, 185) and the Applications alias at (515, 185), Finder's labels
// underneath them. If either script moves those, both move together.

import CoreGraphics
import CoreText
import Foundation
import ImageIO
import UniformTypeIdentifiers

let arguments = CommandLine.arguments
guard arguments.count == 2 else {
    FileHandle.standardError.write(Data("usage: make-dmg-background.swift <output.png>\n".utf8))
    exit(2)
}
let output = URL(fileURLWithPath: arguments[1])

let width: CGFloat = 660
let height: CGFloat = 400
let scale: CGFloat = 2                       // not soft on a Retina display
let appCentre = CGPoint(x: 145, y: 185)      // Finder coordinates, origin top-left
let applicationsCentre = CGPoint(x: 515, y: 185)
let iconHalf: CGFloat = 50

// The site's palette, so the download and the page that offered it look like
// the same product.
func colour(_ hex: UInt32, _ alpha: CGFloat = 1) -> CGColor {
    CGColor(
        srgbRed: CGFloat((hex >> 16) & 0xFF) / 255,
        green: CGFloat((hex >> 8) & 0xFF) / 255,
        blue: CGFloat(hex & 0xFF) / 255,
        alpha: alpha)
}
let paper = colour(0xF6F7F8)
let paperLow = colour(0xE9ECEF)
let ink = colour(0x15191E)
let inkSoft = colour(0x5A6570)
let inkFaint = colour(0x8A939C)
let rule = colour(0xC3CAD1)
let edge = colour(0xD9481F)

guard let context = CGContext(
    data: nil,
    width: Int(width * scale), height: Int(height * scale),
    bitsPerComponent: 8, bytesPerRow: 0,
    space: CGColorSpace(name: CGColorSpace.sRGB)!,
    bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue)
else { fatalError("no drawing context") }

context.scaleBy(x: scale, y: scale)

/// Finder's coordinates run down from the top; CoreGraphics' run up.
func y(_ finderY: CGFloat) -> CGFloat { height - finderY }

// MARK: - Ground

let ground = CGGradient(
    colorsSpace: CGColorSpace(name: CGColorSpace.sRGB),
    colors: [paper, paperLow] as CFArray,
    locations: [0, 1])!
context.drawLinearGradient(
    ground, start: CGPoint(x: 0, y: height), end: CGPoint(x: 0, y: 0), options: [])

// MARK: - Where to drop

/// A soft well behind each icon. It gives the drop target a place, and it is
/// the only thing on the page that is not flat — a card being lifted
/// rather than a decoration.
func well(at centre: CGPoint) {
    let rect = CGRect(
        x: centre.x - 72, y: y(centre.y) - 72, width: 144, height: 144)
    let path = CGPath(roundedRect: rect, cornerWidth: 26, cornerHeight: 26, transform: nil)
    context.saveGState()
    context.setShadow(offset: CGSize(width: 0, height: -3), blur: 14, color: colour(0x15191E, 0.12))
    context.setFillColor(colour(0xFFFFFF))
    context.addPath(path)
    context.fillPath()
    context.restoreGState()
    context.setStrokeColor(colour(0xDDE1E5))
    context.setLineWidth(1)
    context.addPath(path)
    context.strokePath()
}
well(at: appCentre)
well(at: applicationsCentre)

// MARK: - The dimension line

// The site's one ornament, used for the same job: marking a span. Ticked at
// the start, an arrowhead at the end, drawn between the two wells.
let lineY = y(appCentre.y)
let start = appCentre.x + 82
let end = applicationsCentre.x - 82

context.setStrokeColor(rule)
context.setLineWidth(1.5)
context.move(to: CGPoint(x: start, y: lineY))
context.addLine(to: CGPoint(x: end - 10, y: lineY))
context.strokePath()

context.setLineWidth(1.5)
context.move(to: CGPoint(x: start, y: lineY - 7))
context.addLine(to: CGPoint(x: start, y: lineY + 7))
context.strokePath()

context.setFillColor(edge)
context.move(to: CGPoint(x: end, y: lineY))
context.addLine(to: CGPoint(x: end - 14, y: lineY + 8))
context.addLine(to: CGPoint(x: end - 14, y: lineY - 8))
context.closePath()
context.fillPath()

// MARK: - Type

func font(_ name: String, _ size: CGFloat, fallbackWeight: CGFloat = 0) -> CTFont {
    let candidate = CTFontCreateWithName(name as CFString, size, nil)
    // CTFontCreateWithName quietly substitutes when a face is missing, so the
    // check is on what came back rather than on whether a font was returned.
    if (CTFontCopyPostScriptName(candidate) as String) == name { return candidate }
    let system = CTFontCreateUIFontForLanguage(.system, size, nil)!
    guard fallbackWeight != 0 else { return system }
    let traits = [kCTFontWeightTrait: fallbackWeight] as CFDictionary
    let descriptor = CTFontDescriptorCreateCopyWithAttributes(
        CTFontCopyFontDescriptor(system),
        [kCTFontTraitsAttribute: traits] as CFDictionary)
    return CTFontCreateWithFontDescriptor(descriptor, size, nil)
}

enum Alignment { case left, centre }

func draw(
    _ text: String, font: CTFont, colour: CGColor, at point: CGPoint,
    tracking: CGFloat = 0, align: Alignment = .left
) {
    let attributes: [CFString: Any] = [
        kCTFontAttributeName: font,
        kCTForegroundColorAttributeName: colour,
        kCTKernAttributeName: tracking,
    ]
    let line = CTLineCreateWithAttributedString(
        CFAttributedStringCreate(nil, text as CFString, attributes as CFDictionary)!)
    let lineWidth = CTLineGetTypographicBounds(line, nil, nil, nil)
    let x = align == .centre ? point.x - CGFloat(lineWidth) / 2 : point.x
    context.textPosition = CGPoint(x: x, y: y(point.y))
    CTLineDraw(line, context)
}

let mono = font("Menlo-Regular", 9.5)
let monoBold = font("Menlo-Bold", 9.5)
let heading = font("SFPro-Bold", 22, fallbackWeight: 0.4)
let body = font("SFPro-Regular", 12, fallbackWeight: 0)

draw("LATHE", font: monoBold, colour: edge, at: CGPoint(x: 36, y: 44), tracking: 1.8)
draw("Drag onto Applications to install",
     font: heading, colour: ink, at: CGPoint(x: 36, y: 72))

// Over the line, naming what it measures.
draw("DRAG", font: mono, colour: inkFaint,
     at: CGPoint(x: (start + end) / 2, y: appCentre.y - 12), tracking: 2, align: .centre)

// Foot. Where the icons' own labels are not.
draw("Free  ·  Apache-2.0  ·  Runs on your Mac, uploads nothing",
     font: body, colour: inkSoft, at: CGPoint(x: width / 2, y: 330), align: .centre)
draw("lathe  ·  on-device media",
     font: mono, colour: inkFaint, at: CGPoint(x: width / 2, y: 352), tracking: 1.4, align: .centre)

// MARK: - Write

guard let image = context.makeImage() else { fatalError("no image") }
guard let destination = CGImageDestinationCreateWithURL(
    output as CFURL, UTType.png.identifier as CFString, 1, nil)
else { fatalError("cannot write \(output.path)") }

// 144 dpi, so Finder treats the 1320x800 file as a 660x400 image at 2x rather
// than a 1320x800 one at 1x, which would be drawn twice the size and cropped.
CGImageDestinationAddImage(destination, image, [
    kCGImagePropertyDPIWidth: 144,
    kCGImagePropertyDPIHeight: 144,
] as CFDictionary)
guard CGImageDestinationFinalize(destination) else { fatalError("write failed") }
print("    background: \(output.path) (\(Int(width * scale))x\(Int(height * scale)))")
